File: leetcode/cpp/PassedProblems/138-copy-list-with-random-pointer.cpp
/*
 * 138-copy-list-with-random-pointer.cpp
 *
 *  Created on: Jul 14, 2015
 *      Author: hanzhe
 */

#include <iostream>
#include <tr1/unordered_map>
#include <queue>
using namespace std;

/**
 * Definition for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     RandomListNode *next, *random;
 *     RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
 * };
 */
struct RandomListNode {
     int label;
      RandomListNode *next, *random;
      RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
  };
class Solution {
public:
    RandomListNode *copyRandomList(RandomListNode *head) {
    	if(head == NULL)
    		return NULL;
    	std::tr1::unordered_map<RandomListNode *, RandomListNode *> o2n;
    	queue<RandomListNode *> os;
    	std::tr1::unordered_map<RandomListNode *, int> visited;
    	RandomListNode * n1 = new RandomListNode(head->label);
    	os.push(head);
    	o2n[head] = n1;
    	visited[head] = 1;
    	while(os.size() > 0)
    	{
    		RandomListNode * ofr = os.front();
    		RandomListNode * nfr;
    		os.pop();
    		if(o2n.find(ofr) == o2n.end())
    		{
    			nfr = new RandomListNode(ofr->label);
    		}
    		else nfr = o2n[ofr];
    		RandomListNode * onext = ofr->next;
    		if(onext != NULL)
    		{
    			RandomListNode * nnext = o2n.find(onext) == o2n.end() ?
    					new RandomListNode(onext->label) : o2n[onext];
    			o2n[onext] = nnext;
    			if(visited.find(onext) == visited.end())
    			{
    				visited[onext] = 1;
    				os.push(onext);
    			}
    			nfr->next = nnext;
    		}
    		RandomListNode * orand = ofr->random;
    		if(orand != NULL)
    		{
    			RandomListNode * nrand = o2n.find(orand) == o2n.end() ?
    					new RandomListNode(orand->label) : o2n[orand];
    			o2n[orand] = nrand;
    			if(visited.find(orand) == visited.end())
    			{
    				visited[orand] = 1;
    				os.push(orand);
    			}
    			nfr->random = nrand;
    		}
    	}
    	return n1;
    }
};

File: leetcode/python/passedProblems/138-copy-list-with-random-pointer.py
#!/usr/bin/env python
# encoding: utf-8

"""
A linked list is given such that each node contains an additional random pointer which could point to any node in the list or null.

Return a deep copy of the list.
"""


# Definition for singly-linked list with a random pointer.
# class RandomListNode(object):
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

# 网上的方法, 两次过，忘记返回值了
class Solution(object):
    def copyRandomList(self, head):
        mmap = {None:None}
        t = head
        while t != None:
            mmap[t] = RandomListNode(t.label)
            t = t.next
        pre, t = None, head
        while t != None:
            if pre != None: mmap[pre].next = mmap[t]
            mmap[t].random = mmap[t.random]
            pre, t = t, t.next
        return mmap[head]


# 两次过，忘记前后链表了，光顾着random
class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: RandomListNode
        :rtype: RandomListNode
        """
        if head == None: return None
        n2o = {head:RandomListNode(head.label)}
        t = head; npre = None
        while t != None:
            nt = n2o[t] if t in n2o else RandomListNode(t.label)
            if t.random != None and t.random not in n2o:
                n2o[t.random] = RandomListNode(t.random.label)
            if t.random != None:
                nt.random = n2o[t.random]
            if npre != None: npre.next = nt
            t, npre = t.next, nt
        return n2o[head]


